synapse/logging/opentracing.py

```diff
@@ -823,75 +824,110 @@ def extract_text_map(carrier: Dict[str, str]) -> Optional["opentracing.SpanConte
 # Tracing decorators
 
 
-def trace_with_opname(opname: str) -> Callable[[Callable[P, R]], Callable[P, R]]:
+def _create_decorator(
+    func: Callable[P, R],
+    # TODO: What is the correct type for these `Any`? `P.args, P.kwargs` isn't allowed here
+    wrapping_logic: Callable[[Callable[P, R], Any, Any], ContextManager[None]],
+) -> Callable[P, R]:
     """
-    Decorator to trace a function with a custom opname.
-
-    See the module's doc string for usage examples.
-
+    Creates a decorator that is able to handle sync functions, async functions
+    (coroutines), and inlineDeferred from Twisted.
```

Comment:
This logic is generalized from our existing [`trace_with_opname` decorator](https://github.com/matrix-org/synapse/blob/a648a06d52715d0d4ad1ec72d042df1b3fd1be71/synapse/logging/opentracing.py#L826-L886) that handles all of these scenarios.

We now re-use it for both `@trace_with_opname` (which backs `@trace`) and `@tag_args` now ⏩